src/Microsoft.AspNet.PipelineCore/FormFeature.cs

```diff
@@ -1,71 +1,195 @@
 // Copyright (c) Microsoft Open Technologies, Inc. All rights reserved.
 // Licensed under the Apache License, Version 2.0. See License.txt in the project root for license information.
 
+using System;
 using System.Collections.Generic;
 using System.IO;
+using System.Linq;
 using System.Text;
 using System.Threading;
 using System.Threading.Tasks;
-using Microsoft.AspNet.FeatureModel;
 using Microsoft.AspNet.Http;
-using Microsoft.AspNet.HttpFeature;
-using Microsoft.AspNet.PipelineCore.Infrastructure;
+using Microsoft.AspNet.PipelineCore.Collections;
 using Microsoft.AspNet.WebUtilities;
-using Microsoft.AspNet.WebUtilities.Collections;
 
 namespace Microsoft.AspNet.PipelineCore
 {
     public class FormFeature : IFormFeature
     {
-        private readonly IFeatureCollection _features;
-        private readonly FeatureReference<IHttpRequestFeature> _request = FeatureReference<IHttpRequestFeature>.Default;
-        private Stream _bodyStream;
-        private IReadableStringCollection _form;
+        private readonly HttpRequest _request;
 
-        public FormFeature([NotNull] IDictionary<string, string[]> form)
-            : this (new ReadableStringCollection(form))
+        public FormFeature([NotNull] IFormCollection form)
         {
+            Form = form;
         }
 
-        public FormFeature([NotNull] IReadableStringCollection form)
+        public FormFeature([NotNull] HttpRequest request)
         {
-            _form = form;
+            _request = request;
         }
 
-        public FormFeature([NotNull] IFeatureCollection features)
+        public bool HasFormContentType
         {
-            _features = features;
+            get
+            {
+                // Set directly
+                if (Form != null)
+                {
+                    return true;
+                }
+
+                return HasApplicationFormContentType() || HasMultipartFormContentType();
+            }
         }
 
-        public async Task<IReadableStringCollection> GetFormAsync(CancellationToken cancellationToken)
+        public IFormCollection Form { get; set; }
+
+        public IFormCollection ReadForm()
         {
-            if (_features == null)
+            if (Form != null)
             {
-                return _form;
+                return Form;
             }
 
-            var body = _request.Fetch(_features).Body;
+            if (!HasFormContentType)
+            {
+                throw new InvalidOperationException("Incorrect Content-Type: " + _request.ContentType);
+            }
 
-            if (_bodyStream == null || _bodyStream != body)
+            // Read the form async and block, but don't block indefinitely, this could be a 2gb file upload.
+            using (var cts = new CancellationTokenSource(TimeSpan.FromSeconds(30)))
```

Comment:
This would be a really difficult timeout to predict, reproduce, etc in production. How about adding an optional cancellation token argument to ReadForm() but have the default behavior assume a timeout for synchronous-request-body-reading is provided in a different way.